applications/StructuralMechanicsApplication/custom_elements/base_shell_element.cpp

```diff
@@ -0,0 +1,629 @@
+
+    GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        const SizeType index = i * 6;
+        NodeType& i_node = geom[i];
+
+        rResult[index]     = i_node.GetDof(DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = i_node.GetDof(DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = i_node.GetDof(DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = i_node.GetDof(ROTATION_X).EquationId();
+        rResult[index + 4] = i_node.GetDof(ROTATION_Y).EquationId();
+        rResult[index + 5] = i_node.GetDof(ROTATION_Z).EquationId();
+    }
+}
+
+/**
+ * determines the elemental list of DOFs
+ * @param rElementalDofList: the list of DOFs
+ * @param rCurrentProcessInfo: the current process info instance
+ */
+void BaseShellElement::GetDofList(DofsVectorType& rElementalDofList,
+                                  ProcessInfo& rCurrentProcessInfo)
+{
+    rElementalDofList.resize(0);
+    rElementalDofList.reserve(mNumDofs);
+
+    GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        NodeType& i_node = geom[i];
+
+        rElementalDofList.push_back(i_node.pGetDof(DISPLACEMENT_X));
+        rElementalDofList.push_back(i_node.pGetDof(DISPLACEMENT_Y));
+        rElementalDofList.push_back(i_node.pGetDof(DISPLACEMENT_Z));
+
+        rElementalDofList.push_back(i_node.pGetDof(ROTATION_X));
+        rElementalDofList.push_back(i_node.pGetDof(ROTATION_Y));
+        rElementalDofList.push_back(i_node.pGetDof(ROTATION_Z));
+    }
+}
+
+void BaseShellElement::GetValuesVector(Vector& rValues, int Step)
+{
+    if (rValues.size() != mNumDofs)
+        rValues.resize(mNumDofs, false);
+
+    const GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        const NodeType& i_node = geom[i];
+        const array_1d<double, 3>& disp = i_node.FastGetSolutionStepValue(DISPLACEMENT, Step);
+        const array_1d<double, 3>& rot = i_node.FastGetSolutionStepValue(ROTATION, Step);
+
+        const SizeType index = i * 6;
+        rValues[index]     = disp[0];
+        rValues[index + 1] = disp[1];
+        rValues[index + 2] = disp[2];
+
+        rValues[index + 3] = rot[0];
+        rValues[index + 4] = rot[1];
+        rValues[index + 5] = rot[2];
+    }
+}
+
+void BaseShellElement::GetFirstDerivativesVector(Vector& rValues, int Step)
+{
+    if (rValues.size() != mNumDofs)
+        rValues.resize(mNumDofs, false);
+
+    const GeometryType& geom = GetGeometry();
+
+    for (SizeType i = 0; i < geom.size(); ++i)
+    {
+        const NodeType& i_node = geom[i];
+        const array_1d<double, 3>& vel = i_node.FastGetSolutionStepValue(VELOCITY, Step);
+        // TODO also include the angular velocity
+
+        const SizeType index = i * 6;
+        rValues[index]     = vel[0];
+        rValues[index + 1] = vel[1];
+        rValues[index + 2] = vel[2];
+
+        rValues[index + 3] = 0.0;
```

Comment:
rotation_velocity?
If you want the time scheme to work you need to define this

Reply:
I put a todo comment to do this